README.md
# ECMAScript 6 Harmony

ECMAScript 6 Harmony polyfil


## Collections (data structures)

* **Map**<br />
* **Set**<br />

## Array

* **Array.of**<br />
*ECMA-262 6th Edition, 15.4.3.3*<br />

* **Array.from**<br />
*ECMA-262 6th Edition, 15.4.3.4*<br />

## String

* **String.fromCodePoint**<br />
*ECMA-262 6th Edition, 15.5.3.3*<br />

* **String.prototype.codePointAt**<br />
*ECMA-262 6th Edition, 15.5.4.5*<br />

* **String.prototype.repeat**<br />
*ECMA-262 6th Edition, 15.5.4.21*<br />

* **String.prototype.startsWith**<br />
*ECMA-262 6th Edition, 15.5.4.22*<br />

* **String.prototype.endsWith**<br />
*ECMA-262 6th Edition, 15.5.4.23*<br />

* **String.prototype.contains**<br />
*ECMA-262 6th Edition, 15.5.4.24*<br />

* **~~String.prototype.toArray~~**<br />
*~~ECMA-262 6th Edition, 15.5.4.25~~*<br />


## Object

* **~~Object.isObject~~**<br />
*~~ECMA-262 6th Edition, 15.2.3.15~~*<br />

* **Object.getOwnPropertyKeys**<br />
*ECMA-262 6th Edition, 15.2.3.15*<br />

* **Object.is**<br />
*ECMA-262 6th Edition, 15.2.3.16*<br />

* **Object.assign**<br />
*ECMA-262 6th Edition, 15.2.3.17*<br />

* **Object.mixin**<br />
*ECMA-262 6th Edition, 15.2.3.18*<br />


## Number

* **Number.EPSILON**<br />
*ECMA-262 6th Edition, 15.7.3.7*<br />

* **Number.MAX_INTEGER**<br />
*ECMA-262 6th Edition, 15.7.3.7*<br />

* **Number.parseInt**<br />
*ECMA-262 6th Edition, 15.7.3.8*<br />

* **Number.parseFloat**<br />
*ECMA-262 6th Edition, 15.7.3.9*<br />

* **Number.isNaN**<br />
*ECMA-262 6th Edition, 15.7.3.10*<br />

* **Number.isFinite**<br />
*ECMA-262 6th Edition, 15.7.3.11*<br />

* **Number.isInteger**<br />
*ECMA-262 6th Edition, 15.7.3.12*<br />

* **Number.toInteger**<br />
*ECMA-262 6th Edition, 15.7.3.13*<br />

* **Number.prototype.clz**<br />
*ECMA-262 6th Edition, 15.7.3.13*<br />


## Math

* **Math.log10**<br />
*ECMA-262 6th Edition, 15.8.2.19*<br />

* **Math.log2**<br />
*ECMA-262 6th Edition, 15.8.2.20*<br />

* **Math.log1p**<br />
*ECMA-262 6th Edition, 15.8.2.21*<br />

* **Math.expm1**<br />
*ECMA-262 6th Edition, 15.8.2.22*<br />

* **Math.cosh**<br />
*ECMA-262 6th Edition, 15.8.2.23*<br />

* **Math.sinh**<br />
*ECMA-262 6th Edition, 15.8.2.24*<br />

* **Math.tanh**<br />
*ECMA-262 6th Edition, 15.8.2.25*<br />

* **Math.acosh**<br />
*ECMA-262 6th Edition, 15.8.2.26*<br />

* **Math.asinh**<br />
*ECMA-262 6th Edition, 15.8.2.27*<br />

* **Math.atanh**<br />
*ECMA-262 6th Edition, 15.8.2.28*<br />

* **Math.hypot**<br />
*ECMA-262 6th Edition, 15.8.2.29*<br />

* **Math.trunc**<br />
*ECMA-262 6th Edition, 15.8.2.30*<br />

* **Math.sign**<br />
*ECMA-262 6th Edition, 15.8.2.31*<br />

* **Math.cbrt**<br />
*ECMA-262 6th Edition, 15.8.2.32*<br />

* **Math.imul**<br />
*ECMA-262 6th Edition, 15.8.2.33*<br />


## Proposals

* **Object.getOwnPropertyDescriptors**<br />

* **Object.getPropertyDescriptor**<br />

* **Object.getPropertyNames**<br />

* **Object.isnt**<br />

* **Object.isObject**<br />

* **String.prototype.toArray**<br />


## TODO

* *String.prototype.normalize*<br />
* *MapIterator.prototype*<br />
* *RegeExp.prototype...*<br />


## Installation

#### Downloading

Choose one of the ways to download the ECMAScript 6 Harmony polyfil:

*npm*

```
sudo npm install -g es6
```

*git*

```
git clone git@github.com:monolithed/ECMAScript-6.git
```

*zip/gzip*

```
https://github.com/monolithed/ECMAScript-6/downloads
```

#### Server-side including

```javascript
require('./ES6.js');
```

#### Client-side including

```html
<script src="./ES6.js"></script>
```


## Information

NOTE: requires ECMAScript 5! <br />
<br />
[ECMAScript 6 compatibility table] (http://kangax.github.com/es5-compat-table/es6/) <br />
[Historical records of working draft of the ECMA-262 6th specification] (http://wiki.ecmascript.org/doku.php?id=harmony:specification_drafts) <br />
[Proposals of ECMA-262 6th specification] (http://wiki.ecmascript.org/doku.php?id=harmony:proposals) <br />


##.

* ECMAScript 6 Harmony polyfil is licensed under the MIT (MIT_LICENSE.txt) license

* Copyright (c) 2011 [Alexander Guinness] (https://github.com/monolithed)


HISTORY.md
# 0.0.10 / 2013-06-03
* [Critical] Removed String.prototype.toArray, Object.isObject. <br />
ES6 Rev 9 (July 8, 2012 Draft)

# 0.0.9 / 2013-06-02
* [Enhancement] Math.imul. <br />

# 0.0.8 / 2013-06-02
* [Critical] fixed String.prototype.startsWith, String.prototype.endsWith. <br />

# 0.0.7 / 2013-05-19
* [Enhancement] added proposals. <br />

# 0.0.6 / 2013-05-16
* [Enhancement] added Object.getOwnPropertyKeys, Object.mixin, Object.assign. <br />
ES6 Rev 14 (March 8, 2013 Draft)

* [Fix] remove Object.isnt. <br />
ES6 Rev 14 (March 8, 2013 Draft)

# 0.0.5 / 2013-05-16
* [Enhancement] Array.from now takes an optional map function. <br />
ES6 Rev 14 (March 8, 2013 Draft)

# 0.0.4 / 2012-09-20
* [Critical] Fixed String.prototype.startsWith, String.prototype.endsWith

# 0.0.3 / 2012-09-20
* [Critical] Fixed initialization

# 0.0.2 / 2012-08-26
* Added tests
* Fixed Math, Number, Map


tests/Math.js
require('../ES6.js');
require('./suitest.js');

/**
 * ------------------------------------------------------------
 * Math
 * ------------------------------------------------------------
**/


/*
 Math.log10
*/

new Suitest('Math.log10')

.test('Math.log10: > 0', function()
{
	this
		.exec(Math.ceil(Math.log10(10)), 1)
		.done();
})

.test('Math.log10: < 0', function()
{
	this
		.exec(Math.log10(-10), NaN, 'eg')
		.done();
})

.test('Math.log10: NaN', function()
{
	this
		.exec(Math.log10(NaN), NaN, 'eg')
		.done();
})

.test('Math.log10: +0', function()
{
	this
		.exec(Math.log10(0), -Infinity)
		.done();
})

.test('Math.log10: -0', function()
{
	this
		.exec(Math.log10(-0), -Infinity)
		.done();
})

.test('Math.log10: 1', function()
{
	this
		.exec(Math.log10(1), 0)
		.done();
})

.test('Math.log10: +Infinity', function()
{
	this
		.exec(Math.log10(Infinity), Infinity)
		.done();
});


/*
 Math.log2
*/

new Suitest('Math.log2')

.test('Math.log2: > 0', function()
{
	this
		.exec(Math.ceil(Math.log2(10)), 4)
		.done();
})

.test('Math.log2: < 0', function()
{
	this
		.exec(Math.log2(-10), NaN, 'eg')
		.done();
})

.test('Math.log2: NaN', function()
{
	this
		.exec(Math.log2(NaN), NaN, 'eg')
		.done();
})

.test('Math.log2: +0', function()
{
	this
		.exec(Math.log2(0), -Infinity)
		.done();
})

.test('Math.log2: -0', function()
{
	this
		.exec(Math.log2(-0), -Infinity)
		.done();
})

.test('Math.log2: 1', function()
{
	this
		.exec(Math.log2(1), 0)
		.done();
})

.test('Math.log2: +Infinity', function()
{
	this
		.exec(Math.log2(Infinity), Infinity)
		.done();
});


/*
 Math.log1p
*/

new Suitest('Math.log1p')

.test('Math.log1p: > 0', function()
{
	this
		.exec(Math.ceil(Math.log1p(10)), 3)
		.done();
})

.test('Math.log1p: < -1', function()
{
	this
		.exec(Math.log1p(-10), NaN, 'eg')
		.done();
})

.test('Math.log1p: NaN', function()
{
	this
		.exec(Math.log1p(NaN), NaN, 'eg')
		.done();
})

.test('Math.log1p: +0', function()
{
	this
		.exec(Math.log1p(0), 0, 'eg')
		.done();
})

.test('Math.log1p: -0', function()
{
	this
		.exec(Math.log1p(-0), -0, 'eg')
		.done();
})

.test('Math.log1p: +Infinity', function()
{
	this
		.exec(Math.log1p(Infinity), Infinity)
		.done();
});


/*
 Math.expm1
*/

new Suitest('Math.expm1')

.test('Math.expm1: NaN', function()
{
	this
		.exec(Math.expm1(NaN), NaN, 'eg')
		.done();
})

.test('Math.expm1: +0', function()
{
	this
		.exec(Math.expm1(0), 0, 'eg')
		.done();
})

.test('Math.expm1: -0', function()
{
	this
		.exec(Math.expm1(-0), -0, 'eg')
		.done();
})

.test('Math.expm1: +Infinity', function()
{
	this
		.exec(Math.expm1(Infinity), Infinity)
		.done();
})

.test('Math.expm1: -Infinity', function()
{
	this
		.exec(Math.expm1(-Infinity), -1)
		.done();
});


/*
 Math.cosh
*/

new Suitest('Math.cosh')

.test('Math.cosh: NaN', function()
{
	this
		.exec(Math.cosh(NaN), NaN, 'eg')
		.done();
})

.test('Math.cosh: +0', function()
{
	this
		.exec(Math.expm1(0), 0, 'eg')
		.done();
})

.test('Math.cosh: -0', function()
{
	this
		.exec(Math.cosh(-0), -0, 'eg')
		.done();
})

.test('Math.cosh: +Infinity', function()
{
	this
		.exec(Math.cosh(Infinity), Infinity)
		.done();
})

.test('Math.cosh: -Infinity', function()
{
	this
		.exec(Math.cosh(-Infinity), -Infinity)
		.done();
});


/*
 Math.sinh
*/

new Suitest('Math.sinh')

.test('Math.sinh: NaN', function()
{
	this
		.exec(Math.sinh(NaN), NaN, 'eg')
		.done();
})

.test('Math.sinh: +0', function()
{
	this
		.exec(Math.sinh(0), 0, 'eg')
		.done();
})

.test('Math.sinh: -0', function()
{
	this
		.exec(Math.sinh(-0), -0, 'eg')
		.done();
})

.test('Math.sinh: +Infinity', function()
{
	this
		.exec(Math.sinh(Infinity), Infinity)
		.done();
})

.test('Math.sinh: -Infinity', function()
{
	this
		.exec(Math.sinh(-Infinity), -Infinity)
		.done();
});


/*
 Math.tanh
*/

new Suitest('Math.tanh')

.test('Math.tanh: NaN', function()
{
	this
		.exec(Math.tanh(NaN), NaN, 'eg')
		.done();
})

.test('Math.tanh: +0', function()
{
	this
		.exec(Math.tanh(0), 0, 'eg')
		.done();
})

.test('Math.tanh: -0', function()
{
	this
		.exec(Math.tanh(-0), -0, 'eg')
		.done();
})

.test('Math.tanh: +Infinity', function()
{
	this
		.exec(Math.tanh(+Infinity), 1, 'eg')
		.done();
})

.test('Math.tanh: -Infinity', function()
{
	this
		.exec(Math.tanh(-Infinity), -1, 'eg')
		.done();
})

.test('Math.tanh: > 0', function()
{
	this
		.exec(Math.ceil(Math.tanh(10)), 1)
		.done();
})

.test('Math.tanh: < 0', function()
{
	this
		.exec(Math.ceil(Math.tanh(-10)), 0)
		.done();
});


/*
 Math.acosh
*/

new Suitest('Math.acosh')

.test('Math.acosh: NaN', function()
{
	this
		.exec(Math.acosh(NaN), NaN, 'eg')
		.done();
})

.test('Math.acosh: < 1', function()
{
	this
		.exec(Math.acosh(0), NaN, 'eg')
		.done();
})

.test('Math.tanh: 1', function()
{
	this
		.exec(Math.acosh(1), +0, 'eg')
		.done();
})

.test('Math.acosh: +Infinity', function()
{
	this
		.exec(Math.acosh(+Infinity), +Infinity, 'eg')
		.done();
})

.test('Math.acosh: > 0', function()
{
	this
		.exec(Math.ceil(Math.acosh(10)), 3)
		.done();
});


/*
 Math.asinh
*/

new Suitest('Math.asinh')

.test('Math.asinh: NaN', function()
{
	this
		.exec(Math.asinh(NaN), NaN, 'eg')
		.done();
})

.test('Math.asinh: +0', function()
{
	this
		.exec(Math.asinh(0), 0, 'eg')
		.done();
})

.test('Math.asinh: -0', function()
{
	this
		.exec(Math.asinh(-0), -0, 'eg')
		.done();
})

.test('Math.asinh: +Infinity', function()
{
	this
		.exec(Math.asinh(+Infinity), +Infinity, 'eg')
		.done();
})

.test('Math.asinh: -Infinity', function()
{
	this
		.exec(Math.asinh(-Infinity), -Infinity, 'eg')
		.done();
})

.test('Math.asinh: > 0', function()
{
	this
		.exec(Math.ceil(Math.asinh(10)), 3)
		.done();
})

.test('Math.asinh: < 0', function()
{
	this
		.exec(Math.ceil(Math.asinh(-10)), -2)
		.done();
});


/*
 Math.atanh
*/

new Suitest('Math.atanh')

.test('Math.atanh: NaN', function()
{
	this
		.exec(Math.atanh(NaN), NaN, 'eg')
		.done();
})

.test('Math.atanh: < -1', function()
{
	this
		.exec(Math.atanh(-10), NaN, 'eg')
		.done();
})

.test('Math.atanh: > 1', function()
{
	this
		.exec(Math.atanh(10), NaN, 'eg')
		.done();
})

.test('Math.atanh: -1', function()
{
	this
		.exec(Math.atanh(-1), -Infinity, 'eg')
		.done();
})

.test('Math.atanh: +1', function()
{
	this
		.exec(Math.atanh(+1), +Infinity, 'eg')
		.done();
})

.test('Math.atanh: +0', function()
{
	this
		.exec(Math.atanh(0), 0, 'eg')
		.done();
})

.test('Math.atanh: -0', function()
{
	this
		.exec(Math.atanh(-0), -0, 'eg')
		.done();
});


/*
 Math.hypot
*/

new Suitest('Math.hypot')

.test('Math.hypot: +Infinity', function()
{
	this
		.exec(Math.hypot(+Infinity, 1), +Infinity, 'eg')
		.done();
})

.test('Math.hypot: -Infinity', function()
{
	this
		.exec(Math.hypot(-Infinity, 1), -Infinity, 'eg')
		.done();
})

.test('Math.hypot: NaN', function()
{
	this
		.exec(Math.hypot(NaN, 1), NaN, 'eg')
		.done();
})

.test('Math.hypot: +0', function()
{
	this
		.exec(Math.hypot(0, 0), 0, 'eg')
		.done();
})

.test('Math.hypot: -0', function()
{
	this
		.exec(Math.hypot(-0, -0), 0, 'eg')
		.done();
});


/*
 Math.trunc
*/

new Suitest('Math.trunc')

.test('Math.trunc: NaN', function()
{
	this
		.exec(Math.trunc(NaN), NaN, 'eg')
		.done();
})

.test('Math.trunc: +0', function()
{
	this
		.exec(Math.trunc(+0), +0, 'eg')
		.done();
})

.test('Math.trunc: -0', function()
{
	this
		.exec(Math.trunc(-0), -0, 'eg')
		.done();
})

.test('Math.trunc: +Infinity', function()
{
	this
		.exec(Math.trunc(+Infinity), +Infinity, 'eg')
		.done();
})

.test('Math.trunc: -Infinity', function()
{
	this
		.exec(Math.trunc(-Infinity), -Infinity, 'eg')
		.done();
})

.test('Math.trunc: float', function()
{
	this
		.exec(Math.trunc(0.1), 0, 'eg')
		.done();
});


/*
 Math.trunc
*/

new Suitest('Math.sign')

.test('Math.sign: NaN', function()
{
	this
		.exec(Math.sign(NaN), NaN, 'eg')
		.done();
})

.test('Math.sign: +0', function()
{
	this
		.exec(Math.sign(+0), +0, 'eg')
		.done();
})

.test('Math.trunc: -0', function()
{
	this
		.exec(Math.sign(-0), -0, 'eg')
		.done();
})

.test('Math.sign: +1', function()
{
	this
		.exec(Math.sign(+1), +1, 'eg')
		.done();
})

.test('Math.sign: -1', function()
{
	this
		.exec(Math.sign(-1), -1, 'eg')
		.done();
});


/*
 Math.cbrt
*/

new Suitest('Math.cbrt')

.test('Math.cbrt: NaN', function()
{
	this
		.exec(Math.cbrt(NaN), NaN, 'eg')
		.done();
})

.test('Math.cbrt: +0', function()
{
	this
		.exec(Math.cbrt(+0), +0, 'eg')
		.done();
})

.test('Math.cbrt: -0', function()
{
	this
		.exec(Math.cbrt(-0), -0, 'eg')
		.done();
})

.test('Math.cbrt: +Infinity', function()
{
	this
		.exec(Math.cbrt(+Infinity), +Infinity, 'eg')
		.done();
})

.test('Math.cbrt: -Infinity', function()
{
	this
		.exec(Math.cbrt(-Infinity), -Infinity, 'eg')
		.done();
});


/*
 * Math.imul
 * @see https://bugs.webkit.org/show_bug.cgi?id=115143
 */

new Suitest('Math.imul')

.test('Math.imul: Math.imul(1, 0.5)', function()
{
	this
		.exec(Math.imul(1, 0.5), 0)
		.done();
})

.test('Math.imul: Math.imul(1, -0.5)', function()
{
	this
		.exec(Math.imul(1, -0.5), 0)
		.done();
})

.test('Math.imul: Math.imul(2, 1 << 30)', function()
{
	this
		.exec(Math.imul(2, 1 << 30), -2147483648)
		.done();
})

.test('Math.imul: Math.imul(4, 1 << 30)', function()
{
	this
		.exec(Math.imul(4, 1 << 30), 0)
		.done();
})

.test('Math.imul: Math.imul(1, NaN)', function()
{
	this
		.exec(Math.imul(1, NaN), 0)
		.done();
})

.test('Math.imul: Math.imul(1, Infinity)', function()
{
	this
		.exec(Math.imul(1, Infinity), 0)
		.done();
})

.test('Math.imul: Math.imul(0.5, 1)', function()
{
	this
		.exec(Math.imul(0.5, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(-0.5, 1)', function()
{
	this
		.exec(Math.imul(-0.5, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(1 << 30, 2)', function()
{
	this
		.exec(Math.imul(1 << 30, 2), -2147483648)
		.done();
})

.test('Math.imul: Math.imul(1 << 30, 4)', function()
{
	this
		.exec(Math.imul(1 << 30, 4), 0)
		.done();
})

.test('Math.imul: Math.imul(NaN, 1)', function()
{
	this
		.exec(Math.imul(NaN, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(Infinity, 1)', function()
{
	this
		.exec(Math.imul(Infinity, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(NaN, NaN, 2)', function()
{
	this
		.exec(Math.imul(NaN, NaN, 2), 0)
		.done();
})

.test('Math.imul: Math.imul(Infinity, Infinity, 2)', function()
{
	this
		.exec(Math.imul(Infinity, Infinity, 2), 0)
		.done();
})

.test('Math.imul: Math.imul(Infinity, -Infinity)', function()
{
	this
		.exec(Math.imul(Infinity, -Infinity), 0)
		.done();
})

.test('Math.imul: Math.imul(-Infinity, Infinity)', function()
{
	this
		.exec(Math.imul(-Infinity, Infinity), 0)
		.done();
})

.test('Math.imul: Math.imul(-Infinity, -Infinity)', function()
{
	this
		.exec(Math.imul(-Infinity, -Infinity), 0)
		.done();
})

.test('Math.imul: Math.imul(2, 2)', function()
{
	this
		.exec(Math.imul(2, 2), 4)
		.done();
})

.test('Math.imul: Math.imul(2.5, 2)', function()
{
	this
		.exec(Math.imul(2.5, 2), 4)
		.done();
})

.test('Math.imul: Math.imul(2.5, 2)', function()
{
	this
		.exec(Math.imul(2.5, 2), 4)
		.done();
})

.test('Math.imul: Math.imul(2, 2.5)', function()
{
	this
		.exec(Math.imul(2, 2.5), 4)
		.done();
})

.test('Math.imul: Math.imul(2.5, 2)', function()
{
	this
		.exec(Math.imul(2.5, 2), 4)
		.done();
})

.test('Math.imul: Math.imul(2.5, 2.5)', function()
{
	this
		.exec(Math.imul(2.5, 2.5), 4)
		.done();
})

.test('Math.imul: Math.imul(-2, -2)', function()
{
	this
		.exec(Math.imul(-2, -2), 4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, -2)', function()
{
	this
		.exec(Math.imul(-2.5, -2), 4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, -2)', function()
{
	this
		.exec(Math.imul(-2.5, -2), 4)
		.done();
})

.test('Math.imul: Math.imul(-2, -2.5)', function()
{
	this
		.exec(Math.imul(-2, -2.5), 4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, -2)', function()
{
	this
		.exec(Math.imul(-2.5, -2), 4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, -2.5)', function()
{
	this
		.exec(Math.imul(-2.5, -2.5), 4)
		.done();
})

.test('Math.imul: Math.imul(-2, 2)', function()
{
	this
		.exec(Math.imul(-2, 2), -4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, 2)', function()
{
	this
		.exec(Math.imul(-2.5, 2), -4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, 2)', function()
{
	this
		.exec(Math.imul(-2.5, 2), -4)
		.done();
})

.test('Math.imul: Math.imul(-2, 2.5)', function()
{
	this
		.exec(Math.imul(-2, 2.5), -4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, 2)', function()
{
	this
		.exec(Math.imul(-2.5, 2), -4)
		.done();
})

.test('Math.imul: Math.imul(-2.5, 2.5)', function()
{
	this
		.exec(Math.imul(-2.5, 2.5), -4)
		.done();
})

.test('Math.imul: Math.imul(2, -2)', function()
{
	this
		.exec(Math.imul(2, -2), -4)
		.done();
})

.test('Math.imul: Math.imul(2.5, -2)', function()
{
	this
		.exec(Math.imul(2.5, -2), -4)
		.done();
})

.test('Math.imul: Math.imul(2.5, -2)', function()
{
	this
		.exec(Math.imul(2.5, -2), -4)
		.done();
})

.test('Math.imul: Math.imul(2, -2.5)', function()
{
	this
		.exec(Math.imul(2, -2.5), -4)
		.done();
})

.test('Math.imul: Math.imul(2.5, -2)', function()
{
	this
		.exec(Math.imul(2.5, -2), -4)
		.done();
})

.test('Math.imul: Math.imul(2.5, -2.5)', function()
{
	this
		.exec(Math.imul(2.5, -2.5), -4)
		.done();
})

.test('Math.imul: Math.imul(NaN, 1)', function()
{
	this
		.exec(Math.imul(NaN, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(Infinity, 1)', function()
{
	this
		.exec(Math.imul(Infinity, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(1e40, 1)', function()
{
	this
		.exec(Math.imul(1e40, 1), 0)
		.done();
})

.test('Math.imul: Math.imul(0xffffffff, 5)', function()
{
	this
		.exec(Math.imul(0xffffffff, 5), -5)
		.done();
})

.test('Math.imul: Math.imul(0xfffffffe, 5)', function()
{
	this
		.exec(Math.imul(0xfffffffe, 5), -10)
		.done();
})

.test('Math.imul: Math.imul(1, NaN)', function()
{
	this
		.exec(Math.imul(1, NaN), 0)
		.done();
})

.test('Math.imul: Math.imul(1, Infinity)', function()
{
	this
		.exec(Math.imul(1, Infinity), 0)
		.done();
})

.test('Math.imul: Math.imul(1, 1e40)', function()
{
	this
		.exec(Math.imul(1, 1e40), 0)
		.done();
})


tests/Array.js
require('../ES6.js');
require('./suitest.js');

/**
 * ------------------------------------------------------------
 * Array
 * ------------------------------------------------------------
**/

void function(__is__)
{
	/*
	 Array.of
	*/

	new Suitest('Array.of')

	.test('Array.of: returns Array', function()
	{
		this
			.describe("type of Array.of(0, 1) is '[object Array]'")
			.exec(__is__.call(Array.of(0, 1)), '[object Array]')
			.done();
	})

	.test('Array.of: converting a variable number of argument to array', function()
	{
		this
			.describe("Array.of(0, 1) are equal [0, 1]'")
			.exec((Array.of(0, 1).toString()), [0, 1].toString()).done();
	});

	/*
	 Array.from
	*/

	new Suitest('Array.from')

	.test('Array.from: returns Array', function()
	{
		this
			.describe("type of Array.from('01') is [object Array]'")
			.exec(__is__.call(Array.from('01')), '[object Array]').done();
	})

	.test('Array.from: converting a string to array', function()
	{
		this
			.describe("Array.from('01') is [0, 1]'")
			.exec(Array.from('01').toString(), [0, 1].toString()).done();
	})

	.test('Array.from: array from arguments', function()
	{
		this
			.describe("Array.from(arguments) converted to [0, 1]")
			.exec(function() {
				return Array.from(arguments);
		}(0, 1), [0, 1].toString()).done();
	})

	.test('Array.from: [ callback [ value ] ]', function()
	{
		this
			.exec(Array.from('12', function(value) {
				return value;
			}), '1,2').done();
	})

	.test('Array.from: [ callback [ index ] ]', function()
	{
		this
			.exec(Array.from('12', function(value, index) {
				return index;
			}), '0,1').done();
	})

	.test('Array.from: [ callback [ object ] ]', function()
	{
		this
			.exec(Array.from('12', function(value, index, object) {
				return object;
			}), '12,12').done();
	})

	.test('Array.from: [ callback [ value, index ] ]', function()
	{
		this
			.exec(Array.from('12', function(value, index) {
				return value + index;
			}), '10,21').done();
	})

	.test('Array.from: [ callback [ value, index, object ] ]', function()
	{
		this
			.exec(Array.from('12', function(value, index, object) {
				return value + index + object;
			}), '1012,2112').done();
	})

	.test('Array.from: [callback, context ]', function()
	{
		var object = {0: 1, 1: 2};

		this
			.exec(Array.from('01', function(value) {
				return this[value];
			}, object), '1,2').done();
	})

}(Object.prototype.toString);


proposal.js
// -*- coding: utf-8; indent-tabs-mode: nil; tab-width: 4; c-basic-offset: 4; -*-

/**
 * Implementation of ECMAScript 6 (Draft)
 * @requires: ECMAScript 5
 * @author: Alexander Guinness <monolithed@gmail.com>
 * @version:  0.0.1
 * @license:  MIT
 * @date:     Thu Nov 1 00:08:00 2011
 **/


void function(__object__, __array__, __global__)
{
	'use strict';

	var define = function(name)
	{
		var __own__ = __object__.hasOwnProperty;

		if (__own__.call(this, name))
			return 0;

		var set = function(name, value, descriptor)
		{
			Object.defineProperty(this, name, descriptor || {
				value: value,
				configurable: true,
				enumerable:   false,
				writable:     true
			});
		};

		if (__object__.toString.call(name) === '[object Object]')
		{
			for (var key in name) {
				if (__own__.call(name, key))
					set.call(this, key, name[key]);
			}
		}
		else
			set.apply(this, arguments);
	};

	/*
	 * ------------------------------------------------------------
	 *  Object
	 * ------------------------------------------------------------
	 */

	/**
	 * Object.isObject
	 * @edition ECMA-262 6th Edition, 15.2.3.15
	 * Removed in Rev 9 July 8, 2012 ECMA-262 6 Draft
	 *
	 * @param {Object}
	 * @return {Boolean}
	 *
	 * @example:
	 *
	 * Object.isObject({}); // true
	**/
	define.call(Object, 'isObject', function(object) {
		return __object__.toString.call(object) === '[object Object]';
	});


	/**
	 * Object.getOwnPropertyDescriptors
	 * @edition proposal
	 *
	 * Returns a property descriptor of the specified object, including object’s prototype chain
	 * @param {Object} object
	 * Object.getOwnPropertyDescriptor, Array.prototype.forEach
	 * @throws {TypeError}
	 * @return {Object}
	 *
	 * @example:
	 *
	 * var object = {};
	 *
	 * Object.defineProperty(object, 'a', {
	 *   value: 1,
	 *   configurable: false,
	 *   enumerable:   false,
	 *   writable:     false
	 * });
	 *
	 * Object.defineProperty(object, 'b', {
	 *   value:2,
	 *   configurable: true,
	 *   enumerable:   true,
	 *   writable:     true
	 * });
	 *
	 * Object.getOwnPropertyDescriptors(object);
	 *
	 * a: {
	 *   value: 1,
	 *   configurable: false,
	 *   enumerable:   false,
	 *   writable:     false
	 * },
	 *
	 * b: {
	 *   value: 2,
	 *   configurable: true,
	 *   enumerable:   true,
	 *   writable:     true
	 * }
	**/
	define.call(Object, 'getOwnPropertyDescriptors', function(object)
	{
		if (__object__.toString.call(object) !== '[object Object]')
			throw new TypeError('Object.getOwnPropertyDescriptors: ' + object + ' is not an Object!');

		var descriptors = {};

		__array__.forEach.call(Object.getOwnPropertyNames(object), function (property)
		{
			Object.defineProperty(descriptors, property, {
				value: Object.getOwnPropertyDescriptor(object, property),
				configurable: false,
				enumerable:   true,
				writable:     false
			});
		});

		return descriptors;
	});


	/**
	 * Object.getPropertyDescriptor
	 * @edition proposal
	 *
	 * Returns a property descriptor of the specified object, including object’s prototype chain
	 * @param {Object} object
	 * @param {String} name - The name of the property
	 * @throws {TypeError}
	 * @return {Object}
	 *
	 * @example:
	 *
	 * Object.getPropertyDescriptor({}, 'toString');
	 *
	 * {
	 *    value: [Function: toString],
	 *    writable: true,
	 *    enumerable: false,
	 *    configurable: true
	 * }
	**/
	define.call(Object, 'getPropertyDescriptor', function(object, name)
	{
		if (__object__.toString.call(object) !== '[object Object]')
			throw new TypeError('Object.getPropertyDescriptor: ' + object + ' is not an Object!');

		var descriptor = Object.getOwnPropertyDescriptor(object, name),
		__proto__ = Object.getPrototypeOf(object);

		while (descriptor === undefined && __proto__ !== null) {
			descriptor = Object.getOwnPropertyDescriptor(__proto__, name);
			__proto__ = Object.getPrototypeOf(__proto__);
		}

		return descriptor;
	});


	/**
	 * Object.getPropertyNames
	 * Returns an array of all the names of the properties
	 *
	 * @param {Object} object
	 * @throws {TypeError}
	 * @edition proposal
	 * @return {Array}
	 *
	 * @example:
	 *
	 * Object.getPropertyNames({});
	 *
	 * [
	 *  'toString',
	 *  'toLocaleString',
	 *  'hasOwnProperty',
	 *  'valueOf',
	 *  'constructor',
	 *  'propertyIsEnumerable',
	 *  'isPrototypeOf',
	 *  ]
	**/
	define.call(Object, 'getPropertyNames', function(object)
	{
		if (__object__.toString.call(object) !== '[object Object]')
			throw new TypeError('Object.getPropertyNames: ' + object + ' is not an Object!');

		var properies = Object.getOwnPropertyNames(object),
			__proto__ = Object.getPrototypeOf(object);

		while (__proto__ !== null)
		{
			__array__.forEach.call(Object.getOwnPropertyNames(__proto__), function(property) {
				if (properies.indexOf(property) === -1)
					properies.push(property);
			});

			__proto__ = Object.getPrototypeOf(__proto__);
		}

		return properies;
	});


	/**
	 * Object.isnt
	 * Opposed to the Object.isnt
	 *
	 * @param {*} - first generic value for egal comparison
	 * @param {*} - second generic value for egal comparison
	 * @requires Object.is
	 * @return {Boolean}
	 *
	 * @example:
	 *
	 * Object.isnt(0, 0) // false
	**/
	define.call(Object, 'isnt', function(x, y) {
		return !Object.is(x, y);
	});


	/**
	 * String.prototype.toArray
	 * @edition ECMA-262 6th Edition, 15.5.4.25
	 * Removed in Rev 9 July 8, 2012 ECMA-262 6 Draft
	 *
	 * Creates an array from the specified String object
	 * @return {Array} Returns an Array object with elements corresponding to
	 * the characters of this object (converted to a String).
	 *
	 * @example:
	 *
	 * Hello'.toArray() // ['H', 'e', 'l', 'l', 'o'];
	**/
	define.call(String.prototype, 'toArray', function() {
		return this.split('');
	});

}
(Object.prototype, Array.prototype, function() {
	return this;
}());
